### PR TITLE
Macro aware name resoltion

### PR DESCRIPTION
**Pull request by matklad**
The first commit lays the ground work for new name resolution, including

* extracting position-indendent items from parse trees
* walking the tree of modules
* old-style macro_rules resolve

cc @pnkfelix: this looks like an API name resolution should interact with.

### CODE REVIEW

File: crates/ra_mbe/src/mbe_expander.rs [hunk 1]
```diff
@@ -155,7 +155,14 @@ fn match_lhs(pattern: &crate::Subtree, input: &mut TtCursor) -> Result<Bindings,
                 _ => return Err(ExpandError::UnexpectedToken),
             },
             crate::TokenTree::Repeat(crate::Repeat { subtree, kind: _, separator }) => {
+                // Dirty hack to make macro-exapnsion terminate.
```
Comment (bjorn3) [inline crates/ra_mbe/src/mbe_expander.rs]: *macro-expansion

File: crates/ra_mbe/src/mbe_expander.rs [hunk 2]
```diff
@@ -196,7 +203,14 @@ fn expand_tt(
         crate::TokenTree::Repeat(repeat) => {
             let mut token_trees = Vec::new();
             nesting.push(0);
+            // Dirty hack to make macro-exapnsion terminate.
```
Comment (bjorn3) [inline crates/ra_mbe/src/mbe_expander.rs]: *macro-expansion

**Comment by matklad** [comment]
bors try

File: crates/ra_hir/src/nameres/collector.rs [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+    ) -> (PerNs<ModuleDef>, ReachedFixedPoint) {
+        log::debug!("resolving import: {:?} ({:?})", import, self.def_map.edition);
+        if import.is_extern_crate {
+            let res = self.def_map.resolve_name_in_extern_prelude(
+                &import
+                    .path
+                    .as_ident()
+                    .expect("extern crate should have been desugared to one-element path"),
+            );
+            // FIXME: why do we return No here?
```
Comment (flodiebold) [inline crates/ra_hir/src/nameres/collector.rs]: hm, I don't remember if that was intentional... the only reason why the extern prelude might change that I remember is when renaming crates using extern crate, and that is probably not supposed to have an effect for other extern crates...

File: crates/ra_hir/src/nameres/collector.rs [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+            .flat_map(|v| v.iter())
+            .cloned()
+            .collect::<Vec<_>>();
+        for (glob_importing_module, glob_import) in glob_imports {
+            // We pass the glob import so that the tracked import in those modules is that glob import
+            self.update_recursive(glob_importing_module, Some(glob_import), resolutions, depth + 1);
+        }
+    }
+
+    // XXX: this is just a pile of hacks now, because `PerNs` does not handle
```
Comment (flodiebold) [inline crates/ra_hir/src/nameres/collector.rs]: shouldn't we add it there? (although I'm not sure if PerNs is good in general)

Comment (matklad) [reply crates/ra_hir/src/nameres/collector.rs]: I’d rather do this in a separate pr: current impl is enough to check that he approach overlall is sound :)

Comment (matklad) [reply crates/ra_hir/src/nameres/collector.rs]: Although the first step after this PR would probably be patching of impl collection code to account for macro generated impls

**Pull request merged**